File: Done/Binary_Search_307_Questions/55/Solution.py
class Solution:
	def reachNumber(self, target: int) -> int:
		queue = [0]
		step = 1

		while queue:
			curr_level = len(queue)
			for i in range(curr_level):
				current = queue.pop(0)
				go_left = current - step
				go_right = current + step
				
				if go_left == target or go_right == target:
					return step
				
				queue += [go_left, go_right]
			step += 1
		
		return -1

File: Done/Binary_Search_307_Questions/55/Solution.java
class Solution {
    public int reachNumber(int target) {
        target=Math.abs(target);
        int pos=0,step=0,st=1,end=target;

        long sum=0;
        while(st<=end){
            int mid=st+(end-st)/2;
            sum=(long)mid*(mid+1)/2;

            if(sum>=target){
                pos=(int)sum;
                step=mid;
                end=mid-1;
            }
            else{
                st=mid+1;
            }
        }

        while((pos-target)%2!=0){
            step++;
            pos+=step;
        }

        return step;
    }
}